### DISCUSSION TITLE
Profiling in FastAPI

### DISCUSSION THREAD

**Discussion by Partho** [bloomberg/memray#576] [Q&A] [answered]
Hello Memray community,
Any advice on how to integrate Memray with FastAPI? Is there a way we can create a middleware where each API call will be introspected and a Memray stacktrace can be generated in HTML output (like Flamegraph), and shown in the browser?

For instance, in the link: https://bloomberg.github.io/memray/_static/flamegraphs/memray-flamegraph-fib.html
we see the Flamegraph representation in an HTML document. I want to make an API call, (e.g.  GET /foo?profile=1) and it should return the HTML document. I know we can call Memray like this:

``` 
 $ python3 -m memray run -o output.bin my_script.py
 $ python3 -m memray flamegraph output.bin
 ```
but in my case it's an FastAPI application which is up and running in production (we don't have SSH access to production machines hence using memray in the machines via terminal has to be ruled out in my case, and use API call via middleware to retrieve the flamegraph doc). Maybe we can use `subprocess.run` here to run the Memray commands in middleware, but not sure 
what should I replace `my_script.py` with?

Linking a similar call stack profiler for reference -
https://pyinstrument.readthedocs.io/en/latest/guide.html#profile-a-web-request-in-fastapi

**Comment by godlygeek** [comment, marked as answer]
> Is there a way we can create a middleware where each API call will be introspected and a Memray stacktrace can be generated in HTML output (like Flamegraph), and shown in the browser?

Yes, absolutely - you can use [the tracker API](https://bloomberg.github.io/memray/api.html) from a middleware to capture a Memray capture file. We don't have an API for programmatically producing a flame graph from that capture file, but you can easily use `subprocess` for that part, e.g.,
```py
subprocess.check_call([sys.executable, "-m", "memray", "flamegraph", "-o", flamegraph_html_file_name, capture_file])
```
and then you can read that HTML and deliver it to your FastAPI client as a response.
